code/modules/station_goals/dna_vault.dm

```diff
@@ -187,7 +187,7 @@
 		return ..()
 
 /obj/machinery/dna_vault/proc/upgrade(mob/living/carbon/human/H,upgrade_type)
-	if(!(upgrade_type in power_lottery[H]))
+	if(!(upgrade_type in power_lottery[H])||(HAS_TRAIT(H, TRAIT_VIRUSIMMUNE))||(HAS_TRAIT(H, TRAIT_NOBREATH))||(HAS_TRAIT(H, TRAIT_RESISTHEAT))||(HAS_TRAIT(H, TRAIT_NOFIRE))||(HAS_TRAIT(H, TRAIT_PIERCEIMMUNE))||(H.next_move_modifier == 0.5)||(H.has_movespeed_modifier(/datum/movespeed_modifier/dna_vault_speedup))||(H.dna.species.armor==30)||(H.dna.species.stunmod==0.5))
```

Comment:
this is a very bad way to check if the user has one of the upgrades. people can receive or change almost all of these things you check for from other sources and then theyll be excluded

Reply:
suggest an alternative, it doesn't give a vault specific trait